wireless_modules/mqtt_test/umqttsimple.py

```diff
@@ -0,0 +1,207 @@
+try:
+    import usocket as socket
+except:
+    import socket
+import ustruct as struct
+from ubinascii import hexlify
+
+class MQTTException(Exception):
+    pass
+
+class MQTTClient:
```

Comment:
Is this just the paho-mqtt library code that's been replicated here? One question I have is, what do we do when the paho-mqtt library code is updated? Do we need to update this manually? Can we automate this somehow?

Reply:
There is no official 'paho-mqtt' library for MicroPython. I assume because of the limited space requirements and the fact that paho is meant to have all of the features of MQTT. The MicroPython `umqttsimple.py` script is a slimmed down client with all the key features we need. 

About updating, I'm really not sure how this would work as I literally just copied from the repo to get this to work. Because the MQTT standard does not change often, I can't imagine we would have to update the script regularly. 

I am very keen for input however, as I feel like it is a bit of a hack.

[This it the umqttsimple repo if you want to check it out](https://github.com/RuiSantosdotme/ESP-MicroPython/blob/master/code/MQTT/umqttsimple.py)

[Alternative MQTT implementation](https://github.com/micropython/micropython-lib/tree/master/umqtt.simple)

Reply:
Okay this is actually a really easy fix.

We can just use upip, which is micro-pip. It will just check whether the latest version of MQTT is installed and install the latest version automatically.